### PR TITLE
Do not cache DNS queries that have timed out

### PR DESCRIPTION
**Pull request by trustin**
Motivation:

`CachingDnsResolver` should not cache any DNS queries that have timed out. However, it currently caches them when the timeout exception is buried in the exception's root cause or suppressed causes.

Modifications:

- Added `DnsUtil.isDnsQueryTimedOut()` to determine whether the DNS query failure was due to a timeout or not precisely
- Changed `DnsTimeoutUtil.assertDnsTimeoutException()` to use `DnsUtil.isDnsQueryTimedOut()`.
- Changed `CachingDnsResolver` to use `DnsUtil.isDnsQueryTimedOut()` in addition to the `instanceof UnknownHostException`.

Result:

- Armeria's DNS resolver doesn't cache a DNS query timeout.

### CODE REVIEW

**Review by jrhee17 — APPROVED** [review]
👍 👍 👍

File: core/src/main/java/com/linecorp/armeria/internal/client/dns/CachingDnsResolver.java [hunk 1]
```diff
@@ -78,20 +78,29 @@ private CompletableFuture<List<DnsRecord>> resolve0(DnsQuestionContext ctx, DnsQ
...
+                                if (DnsUtil.isDnsQueryTimedOut(cause)) {
+                                    logger.debug("[{}] Not caching a timed-out DNS query: {}",
+                                                 name, question);
+                                } else {
+                                    logger.debug("[{}] Caching a failed DNS query: {}, cause: {}",
+                                                 name, question, cause.getMessage());
+                                    dnsCache.cache(key, (UnknownHostException) cause);
+                                }
+                            } else {
+                                logger.debug("[{}] Not caching a DNS query failed with an unexpected cause: {}",
```
[the remark this answers is not in the document]
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/internal/client/dns/CachingDnsResolver.java:102]: That's correct indeed. Let me simplify logging.

[the remark this answers is not in the document]
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/internal/client/dns/CachingDnsResolver.java]: e4a7cee624809bb104205d731fcd05a99cacb21e

**Comment by trustin** [comment]
Thanks a lot for quick reviews, folks. :bow:

**Pull request merged**